File: 11.6.5.c
#include <stdio.h>
#include <string.h>
#include <limits.h>

#define MAXCUTS 50

int memo[MAXCUTS+1][MAXCUTS+1];
int cuts[MAXCUTS+1];

int min_cost(int L, int R, int i, int j) {
  if (memo[i][j] != -1) {
    return memo[i][j];
  }
  if (i == j) {
    memo[i][j] = 0;
    return 0;
  }

  int min = INT_MAX;
  int k, cost;

  for (k = i; k < j; k++) {
    cost = (R - L) + min_cost(L, cuts[k], i, k) + min_cost(cuts[k], R, k + 1, j);
    if (cost < min) {
      min = cost;
    }
  }

  memo[i][j] = min;
  return min;
}

int main() {
  int stick_length, num_cuts, i, j;

  while((scanf("%d", &stick_length) != EOF) && stick_length != 0) {
    scanf("%d", &num_cuts);
    for (i = 0; i < num_cuts; i++){
      scanf("%d", &cuts[i]);
    }
    for (i = 0; i < MAXCUTS; i++) {
      for (j = 0; j < MAXCUTS; j++) {
        memo[i][j] = -1;
      }
    }
    printf("The minimum cutting is %d.\n", min_cost(0, stick_length, 0, num_cuts));
  }

  return 0;
}

File: 11.6.5.py
import sys

MAX_STICK_LENGTH = 1000
memo = {}


def minimum_cutting_cost(L, R, cut_positions):
    if (L, R) in memo:
        return memo[(L, R)]
    if not cut_positions:
        memo[(L, R)] = 0
        return 0

    minimum_cost = sys.maxsize
    for i, cut_position in enumerate(cut_positions):
        cost = (R - L) + minimum_cutting_cost(
            L, cut_position, cut_positions[:i]) + minimum_cutting_cost(
                cut_position, R, cut_positions[i + 1:])

        if cost < minimum_cost:
            minimum_cost = cost

    memo[(L, R)] = minimum_cost
    return minimum_cost


for line in sys.stdin:
    stick_length = int(line)
    if stick_length == 0:
        break
    num_cuts = int(sys.stdin.readline())
    cut_positions = [int(x) for x in sys.stdin.readline().split()]

    memo = {}
    print(
        f'The minimum cutting is {minimum_cutting_cost(0, stick_length, cut_positions)}.'
    )
